### DISCUSSION TITLE
Differentiate MONEY entities

### DISCUSSION THREAD

**Discussion by wagnertail** [explosion/spaCy#9434] [Help: Coding & Implementations] [answered]
I want to differentiate MONEY values using NER (and possibly POS tagging combined with the dependency parser) - is this possible?  Do I need to list all verbs that are 'related' to each type of 'MONEY' value?  

I would like to differentiate expenditure from income, i.e.:

1. I earned £10,000
2.  I spent £200

so that I can create entity labels, eg: £10,000: MONEY_INCOME, £200: MONEY_EXPENSE

SpaCy documentation does touch on this though it focuses on verb tenses, which doesn't help me with the above query (the example is 'Using entities, part-of-speech tags and the dependency parse' under Rule-based Matching/ Models & Rules) 

Any help much appreciated!  Thanks!

**Comment by DuyguA** [comment, marked as answer]
Hellos,

I'm not sure you need training, can't we just hunt down the money entities with the patterns:

```
currency_symbol number
number currency_symbol
```

Coming to whether the money is spent, earned, lost, stolen, inherited and so on, it depends on the sentence semantics. This is a sentence level semantical task rather than parsing. But you can still play some tricks here. One can find the verb that the money entity attached in the dependency parser and examine the verb a bit. I can give some clues if you wish :wink:

**Reply by wagnertail** [reply]
Thanks so much DuyguA - any clues much appreciated! 🙏

**Reply by DuyguA** [reply]
OK, let's start easy and extract the money entities first:

```
import spacy
nlp = spacy.load("en_core_web_md")
from spacy.matcher import Matcher
doc1 = nlp("I spent 200$ on my books.")
matcher = Matcher(nlp.vocab)
pattern = [{"IS_DIGIT": True}, {"IS_CURRENCY": True}]
matcher.add("money", [pattern])
matches = matcher(doc1)
for mid, start, end in matches:
    print(start, end, doc1[start:end])
... 
2 4 200$
```

Here `pattern` matches a sequence of 2 tokens, a number and a currency symbol. Some currencies are written before the amount, so I'll add another pattern which is `currency` , `number`:

```
pattern2 = [{"IS_CURRENCY": True}, {"IS_DIGIT": True}]
matcher.add("money2", [pattern2])
doc2 =nlp("Your flight costs £20. You can buy extra leg space for £5")
matches = matcher(doc2)
for mid, start, end in matches:
...    print(start, end, doc2[start:end])
... 
3 5 £20
13 15 £5
```

I hope matching part is clear. No we can walk on the dependency tree :wink:

**Comment by DuyguA** [comment]
Let's print first sentence's dependency tree to have closer look:

![image](https://raw.githubusercontent.com/DuyguA/spacy_helpers/master//tre_200%24.png)

There are some points:

*  Head of `200` is `$` . 
*  `200` is a numerical modifier to `$` (relation is `nummod`)
*  `$`'s head is `spent`.
* `$` attaches to `spent` with relation `dobj` because `200$` is direct object of this verb. 

Basically we follow the dependency arcs `200` -> `$` -> `spent` to hunt down the verb related to this money amount.
Another example is a passive construction, the dependency relation is different but we follow the arcs in same way:

 ![image](https://raw.githubusercontent.com/DuyguA/spacy_helpers/master//200passive.png)

Here `200$` is a passive subject of the verb `stolen` . Still you can follow the arcs 200 -> $ -> stolen. Let's do it with code:

```
doc3 = nlp("200$ is stolen from my account")
matches = matcher(doc3)
for mid,start,end in matches:
...    match = doc3[start:end]
...    curr_tok = match[-1]
...    print(curr_tok.dep_, curr_tok.head, curr_tok.head.lemma_)
... 
nsubjpass stolen steal
```
Now we detected the verb attached to money amount. 3rd part is about verb semantics, keep on reading :+1:

**Comment by DuyguA** [comment]
OK, 3rd part is about the verb and sentence semantics. Some verbs gives away the sentence semantics, even if you don't see other context words such as:

```
earn
spend
steal
win
lose
charge
pay
```
Some verbs are not so easy:

```
get 
take
```

Some examples:

Bank charged 200$ me for getting a credit card.
I spent 200$ on my new books. 
He earns around 500$ from his part time job.
 
 For this verbs, you can extract the verb and make a lookup with a predefined list of verbs most probably.
 
These ones are not easy:

```
200$ is taken from my account by tax office.   (you paid money for tax)
200$ is taken from my account without my permission.  (your money is stolen)
```
For this verbs you might need a tiny classifier for the sentence semantics.

**Reply by wagnertail** [reply]
Thanks so much Duygu - this is excellent, very kind of to show me the process 🙏🙏🙏🙏

**Reply by DuyguA** [reply]
> Thanks so much Duygu - this is excellent, very kind of to show me the process 

If you want more examples, I can offer some chapters from my newly published book. Matcher section has lot of entity extraction examples:
https://github.com/PacktPublishing/Mastering-spaCy/tree/main/Chapter04

There's another section which offers  what you asked: how to extract entities, their attached verbs and play tricks on them.  I made examples on ATIS dataset. This is the chapter: https://github.com/PacktPublishing/Mastering-spaCy/tree/main/Chapter06


I'm happy to help too, hope to see you in new discussions! :heart:
